src/main/java/org/opensearch/neuralsearch/processor/normalization/MinMaxScoreNormalizationTechnique.java

```diff
@@ -220,47 +257,78 @@ private float[] getMinScores(final List<CompoundTopDocs> queryTopDocs, final int
         return minScores;
     }
 
-    private float normalizeSingleScore(final float score, final float minScore, final float maxScore, final LowerBound lowerBound) {
+    private float normalizeSingleScore(
+        final float score,
+        final float minScore,
+        final float maxScore,
+        final LowerBound lowerBound,
+        final UpperBound upperBound
+    ) {
         // edge case when there is only one score and min and max scores are same
-        if (Floats.compare(maxScore, minScore) == 0 && Floats.compare(maxScore, score) == 0) {
+        if (isSingleScore(score, minScore, maxScore)) {
             return SINGLE_RESULT_SCORE;
         }
-        if (lowerBound.isEnabled() == false) {
-            return LowerBound.Mode.IGNORE.normalize(score, minScore, maxScore, lowerBound.getMinScore());
+
+        float effectiveMinScore = lowerBound.determineEffectiveScore(score, minScore, maxScore);
+        float effectiveMaxScore = upperBound.determineEffectiveScore(score, minScore, maxScore);
+
+        if (lowerBound.shouldClipToBound(score, effectiveMinScore)) {
+            return MIN_SCORE;
+        }
+        if (upperBound.shouldClipToBound(score, effectiveMaxScore)) {
+            return MAX_SCORE;
         }
-        return lowerBound.getMode().normalize(score, minScore, maxScore, lowerBound.getMinScore());
+
+        return calculateNormalizedScore(score, effectiveMinScore, effectiveMaxScore);
     }
 
-    /**
-     * Get lower bounds from input parameters
-     * @param params user provided input parameters for this technique
-     * @return optional list of lower bounds. Can be empty in case lower bounds are not provided
-     */
-    private Optional<List<Pair<LowerBound.Mode, Float>>> getLowerBounds(final Map<String, Object> params) {
+    private boolean isSingleScore(float score, float minScore, float maxScore) {
+        return Floats.compare(maxScore, minScore) == 0 && Floats.compare(maxScore, score) == 0;
+    }
+
+    private float calculateNormalizedScore(float score, float effectiveMinScore, float effectiveMaxScore) {
+        float normalizedScore = (score - effectiveMinScore) / (effectiveMaxScore - effectiveMinScore);
```

Comment:
should we add protection from division by zero. In theory scenario when effectiveMaxScore == effectiveMinScore is possible. Simple check will do the job:
```
    if (Floats.compare(effectiveMaxScore, effectiveMinScore) == 0) {
        return SINGLE_RESULT_SCORE;
    }
```

Reply:
@ryanbogan  where did we addressed this concern?

Reply:
also can you add a test specifically for scenario when effectiveMaxScore == effectiveMinScore

Reply:
sounds good